opm/simulators/wells/MultisegmentWell_impl.hpp

```diff
@@ -566,24 +566,112 @@ namespace Opm
+        initPrimaryVariablesEvaluation();
+
+        // get the bhp value based on the bhp constraints
+        const double bhp = Base::mostStrictBhpFromBhpLimits(deferred_logger);
+
+        // does the well have a THP related constraint?
+        if ( !Base::wellHasTHPConstraints() ) {
+            assert(std::abs(bhp) != std::numeric_limits<double>::max());
+
+            computeWellRatesWithBhp(ebosSimulator, bhp, /*iterate=*/ true, well_potentials, deferred_logger);
+        } else {
+
+            const std::string msg = std::string("Well potential calculation is not supported for thp controlled multisegment wells \n")
+                    + "A well potential of zero is returned for output purposes. \n"
+                    + "If you need well potential computed from thp to set the guide rate for group controled wells \n"
+                    + "you will have to change the " + name() + " well to a standard well \n";
+
+            deferred_logger.warning("WELL_POTENTIAL_FOR_THP_NOT_IMPLEMENTED_FOR_MULTISEG_WELLS", msg);
+            return;
+        }
+
     }
 
 
+    template<typename TypeTag>
+    void
+    MultisegmentWell<TypeTag>::
+    computeWellRatesWithBhp(const Simulator& ebosSimulator,
+                            const double& bhp,
+                            const bool iterate,
+                            std::vector<double>& well_flux,
+                            Opm::DeferredLogger& deferred_logger)
+    {
+        // store a copy of the well state, we don't want to update the real well state
+        WellState copy = ebosSimulator.problem().wellModel().wellState();
+
+        // try to improve the initial condition.
+        // we adjust the initial seg pressure with half the difference
+        // between the real bhp and the potential bhp.
+        // This is ad-hoc, but seems to work reasonbly well.
+        const double adjustBhp = (bhp - primary_variables_[0][SPres]);
+        double adjustTotalRate = 0.0;
+        for (int p = 0; p < number_of_phases_; ++p) {
+            adjustTotalRate += copy.wellPotentials()[index_of_well_ * number_of_phases_ + p];
+        }
+
+        if (primary_variables_[0][GTotal] > 0)
+            adjustTotalRate /= primary_variables_[0][GTotal];
+
+        for (int seg = 0; seg < numberOfSegments(); ++seg) {
+            primary_variables_[seg][SPres] += adjustBhp;
+            primary_variables_[seg][SPres] = Opm::max(primary_variables_[seg][SPres], 1e5);
+            primary_variables_[seg][GTotal] *= adjustTotalRate;
+        }
+
+        initPrimaryVariablesEvaluation();
+
+        if (iterate) {
+            const double dt = ebosSimulator.timeStepSize();
+            // iterate to get a solution that satisfies the bhp potential.
+            iterateWellEquations(ebosSimulator, dt, copy, deferred_logger);
+        }
+
+        // compute the potential and store in the flux vector.
+        const int np = number_of_phases_;
```

Comment:
I believe the following code will be equivalent to get the well rates based on `getSegmentRate(0, comp_idx)` . Basically the well rates of the top segment should be same with summing through all the perforation rates. 

Please report if they turned out to be different.